File: 3439-reschedule-meetings-for-maximum-free-time-i/3439-reschedule-meetings-for-maximum-free-time-i.cpp
class Solution {
public:
  int maxFreeTime(int eventTime, int k, const vector<int>& startTime, const vector<int>& endTime) {
    int n = startTime.size();
    int windowSum = 0, maxFree = 0;

    // Helper lambda to compute the gap at position i
    auto getGap = [&](int i) -> int {
      if (i == 0) return startTime[0];  // Before first event
      if (i == n) return eventTime - endTime.back();  // After last event
      return startTime[i] - endTime[i - 1];  // Between events
    };

    // Step 1: Initial window sum of size k+1
    for (int i = 0; i <= k && i <= n; ++i)
      windowSum += getGap(i);
    maxFree = windowSum;

    // Step 2: Slide the window
    for (int i = k + 1; i <= n; ++i) {
      windowSum += getGap(i) - getGap(i - k - 1);
      maxFree = max(maxFree, windowSum);
    }

    return maxFree;
  }
};

File: 3439-reschedule-meetings-for-maximum-free-time-i/3439-reschedule-meetings-for-maximum-free-time-i.py
class Solution:
    def maxFreeTime(self, eventTime: int, k: int, startTime: list[int], endTime: list[int]) -> int:
        n = len(startTime)

        # Precompute gaps array (faster than computing inside loop)
        gaps = [0] * (n + 1)
        gaps[0] = startTime[0]  # Before first event

        for i in range(1, n):
            gaps[i] = startTime[i] - endTime[i - 1]  # Between events

        gaps[n] = eventTime - endTime[-1]  # After last event

        # Sliding window
        window_sum = sum(gaps[:k + 1])
        max_free = window_sum

        for i in range(k + 1, len(gaps)):
            window_sum += gaps[i] - gaps[i - k - 1]
            max_free = max(max_free, window_sum)

        return max_free
